File: DSA/Asmi_Garg/day4_ques2.java
import java.util.Scanner;

public class SameFrequencyChecker {

    public static String isValid(String s) {
        int[] charCount = new int[26];
        for (char c : s.toCharArray()) {
            charCount[c - 'a']++;
        }

        int[] frequencyCount = new int[s.length() + 1];
        for (int count : charCount) {
            if (count > 0) {
                frequencyCount[count]++;
            }
        }

        int uniqueFreqCount = 0;
        int[] uniqueFreq = new int[2];
        int[] freqOccurrences = new int[2];

        for (int i = 1; i < frequencyCount.length; i++) {
            if (frequencyCount[i] > 0) {
                uniqueFreq[uniqueFreqCount] = i;
                freqOccurrences[uniqueFreqCount] = frequencyCount[i];
                uniqueFreqCount++;
            }
        }

        if (uniqueFreqCount == 1) {
            return "valid";
        }

        if (uniqueFreqCount == 2) {
            int freq1 = uniqueFreq[0];
            int freq2 = uniqueFreq[1];
            int count1 = freqOccurrences[0];
            int count2 = freqOccurrences[1];

            if ((freq1 == 1 && count1 == 1) || (freq2 == 1 && count2 == 1)) {
                return "valid";
            }

            if ((Math.abs(freq1 - freq2) == 1) && ((freq1 > freq2 && count1 == 1) || (freq2 > freq1 && count2 == 1))) {
                return "valid";
            }
        }

        return "invalid";
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the string: ");
        String input = scanner.nextLine();
        System.out.println(isValid(input));
        scanner.close();
    }
}

File: DSA/Umang-Singhal/Day4_Ques2.c
#include <stdio.h>
#include <string.h>
int main() {
    char str[100];
    int freq[26] = {0};
    int len, i, j, unique_freq[26] = {0}, unique_count = 0;
    printf("Enter the string: ");
    scanf("%s", str);   
    len = strlen(str);
    for (i = 0; i < len; i++) {
        freq[str[i] - 'a']++;
    }
    for (i = 0; i < 26; i++) {
        if (freq[i] > 0) {
            int found = 0;
            for (j = 0; j < unique_count; j++) {
                if (unique_freq[j] == freq[i]) {
                    found = 1;
                    break;
                }
            }
            if (!found) {
                unique_freq[unique_count] = freq[i];
                unique_count++;
            }
        }
    }
    if (unique_count == 1) {
        printf("valid\n");
    } else if (unique_count == 2) {
        int count1 = 0, count2 = 0;
        for (i = 0; i < 26; i++) {
            if (freq[i] == unique_freq[0]) count1++;
            else if (freq[i] == unique_freq[1]) count2++;
        }
        if ((count1 == 1 && (unique_freq[0] == 1 || unique_freq[0] - 1 == unique_freq[1])) || 
            (count2 == 1 && (unique_freq[1] == 1 || unique_freq[1] - 1 == unique_freq[0]))) {
            printf("valid\n");
        } else {
            printf("invalid\n");
        }
    } else {
        printf("invalid\n");
    }
    return 0;
}
